### ISSUE TITLE
torch.topk sorted parameter

### ISSUE THREAD

**Issue by RohitKeshari** [pytorch/pytorch#3982]
In the documentation **torch.topk** uses four parameter such as topk(self, k, dim=None, largest=True, sorted=True). In case of **sorted=False**, it gives ascending order sorted elements. This is quite confusing. If **sorted=False** it should provide first come first serve elements. Essentially index should be sorted.

**Comment by soumith** [comment]
the function is `topk`, it will return the topk elements. The kwarg `sorted` should've been named `descending`

**Comment by RohitKeshari** [comment]
Thanks for your reply @soumith. The top k elements might be in form of ascending, descending, and without sorted (based on smalles index first). It would be better if `topk` can have all these functionalities.

**Comment by colesbury** [comment]
@soumith, that's not correct. If `sorted` is False, the output ordering is **unspecified**. It's not always descending -- that was just an artifact of the input data.

**Comment by colesbury** [comment]
For example:

```
>>> torch.randn(10000).topk(10, sorted=False)
 3.1205
 3.0569
 3.0515
 3.0308
 3.2317
 3.3175
 3.4943
 3.5087
 4.0495
 3.8342
[torch.FloatTensor of size 10]
```

The reason the **sorted** parameter exists is that it's faster to compute the top-K elements than it is to compute and sort the top-k elements.

**Comment by RohitKeshari** [comment]
Thanks for clarification @colesbury, do you think if the output of the `topk` is based on ascending, descending, and without sorted (based on smallest index first) is a good idea. I think **unspecified** output might not useful for the user because of the uncertainty of the output.

**Comment by apaszke** [comment]
If unspecified makes the algorithm a lot faster, it might still be desirable. Consider a case of computing top-k error rates on some classification problem. You take the indices of max logits, and it doesn’t really matter for you if they are sorted or not, you will have to iterate over all of them anyway.

**Comment by RohitKeshari** [comment]
I get the idea of having uspecified output. However it is not metioned in the documentation of the `topk`. Can you also suggest me the better way of having top k elements without sorted (based on smallest index first). Since `topk` also return indexes of the elements, I can sort the indexes and have it from the main _vector_. But it introduces extra overhead while I already have top k elements.

**Comment by ghrua** [comment]
@RohitKeshari If you want to get get the top k elements sorted by its index, you could do this:

```
In [1]: import torch
In [2]: x = torch.rand(3, 5)
In   [3]: val, idx = torch.topk(x, k=3, dim=-1, largest=False) # k smallest elements
In   [4]: sorted_idx, new_idx = torch.sort(idx, dim=-1)
In   [5]: val = torch.gather(val, dim=-1, index=new_idx)
In   [6]: x
Out[6]:

 0.1868  0.7928  0.5790  0.4628  0.2190
 0.1212  0.9903  0.4272  0.1662  0.8974
 0.1360  0.2362  0.5849  0.5407  0.6900
[torch.FloatTensor of size 3x5]

In   [7]: val
Out[7]:

 0.1868  0.4628  0.2190
 0.1212  0.4272  0.1662
 0.1360  0.2362  0.5407
[torch.FloatTensor of size 3x3]

In   [8]: sorted_idx
Out[8]:

 0  3  4
 0  2  3
 0  1  3
[torch.LongTensor of size 3x3]

```
